File: 数据结构与算法/排序/sort.js
/*

排序算法

*/

/**
 * 生成一个随机整数组
 * @param {number} len 数组长度
 * @param {number} left 随机整数的下界范围
 * @param {number} right 随机整数的上界范围
 * @returns {number[]} 一个生成的随机整数组
 */
function genRandVec(len, left, right) {
    let vec = [];
    while (len > 0) {
        vec.push(parseInt(Math.random() * (right - left + 1) + left));
        len--;
    }
    return vec;
}

/**
 * 输出一个数组
 * @param {Array} vec 待输出数组
 * @param {string} prefix 输出前缀
 */
function printVec(vec, prefix) {
    process.stdout.write(prefix + ":");
    for (const e of vec)
        process.stdout.write(' ' + e);
    console.log();
}

// /*********************************************************************************************************************/

// 插入类排序

/**
 * 直接插入排序
 * @param {Array} vec 待排序数组
 */
// 时间复杂度最好O(n)，最坏O(n^2)，平均O(n^2)
// 空间复杂度O(1)
// 稳定
function insertSort(vec) {
    for (let i = 1; i < vec.length; i++) {
        let tmp = vec[i];
        let idx = i - 1;
        for (; idx >= 0 && vec[idx] > tmp; idx--)
            vec[idx + 1] = vec[idx];
        vec[idx + 1] = tmp;
    }
}

/**
 * 从数组的顺序部分[left,right]寻找下一个元素要被插入的位置
 * @param {Array} vec 待寻找数组
 * @param {number} left 寻找范围左闭边界索引
 * @param {number} right 寻找范围右闭边界索引
 * @param {any} target 待插入的元素
 * @returns {number} 待插入索引
 */
function bindex(vec, left, right, target) {
    let mid = parseInt((left + right) / 2);
    while (left < right) {
        if (vec[mid] === target)
            return mid + 1;
        else if (vec[mid] < target)
            left = mid + 1;
        else
            right = mid - 1;
        mid = parseInt((left + right) / 2);
    }
    return vec[mid] <= target ? mid + 1 : mid;
}

/**
 * 折半插入排序
 * @param {Array} vec 待排序数组
 */
// 时间复杂度最好O(nlog2n)，最坏O(n^2)，平均O(n^2)
// 空间复杂度O(1)
// 稳定
function binsertSort(vec) {
    for (let i = 1; i < vec.length; i++) {
        let index = bindex(vec, 0, i - 1, vec[i]);
        let tmp = vec[i];
        let idx = i - 1;
        for (; idx >= index; idx--)
            vec[idx + 1] = vec[idx];
        vec[idx + 1] = tmp;
    }
}

/**
 * 希尔排序
 * @param {Array} vec 待排序数组
 * @param {number} k 初始增量
 * @param {Function} kfun 增量选取函数
 */
// 时间复杂度和增量选取有关，平均O(nlog2n)
// 空间复杂度O(1)
// 不稳定
function shellSort(vec, k, kfun) {
    while (k > 1) {
        for (let i = 0; i < k; i++)
            for (let j = i + k; j < vec.length; j += k) {
                let tmp = vec[j];
                let idx = j - k;
                for (; idx >= i && vec[idx] > tmp; idx -= k) // 这里注意要和vec[j]即tmp比较，为它寻找一个插入的地方，而不是和vec[idx+1]比较，会出错
                    vec[idx + k] = vec[idx];
                vec[idx + k] = tmp;
            }
        k = kfun(k);
    }
    for (let i = 1; i < vec.length; i++) {
        let tmp = vec[i];
        let idx = i - 1;
        for (; idx >= 0 && vec[idx] > tmp; idx--) // 同上
            vec[idx + 1] = vec[idx];
        vec[idx + 1] = tmp;
    }
}

// 交换类排序

/**
 * 冒泡排序
 * @param {Array} vec 待排序数组
 */
// 时间复杂度最好O(n)，最坏O(n^2)，平均O(n^2)
// 空间复杂度O(1)
// 稳定
function bubbleSort(vec) {
    for (let i = vec.length - 1; i >= 1; i--) {
        let flag = false;
        for (let j = 0; j < i; j++)
            if (vec[j] > vec[j + 1]) {
                [vec[j], vec[j + 1]] = [vec[j + 1], vec[j]];
                flag = true;
            }
        if (!flag)
            return;
    }
}

/**
 * 快速排序
 * @param {Array} vec 待排序数组
 * @param {number} left 待排序部分的左闭区间边界索引
 * @param {number} right 待排序部分的右闭区间边界索引
 */
// 时间复杂度最好O(nlog2n)，最坏O(n^2)，数据越无序越好，越有序越差，在所有O(nlog2n)的算法中基本操作次数多项式最高次项最小
// 空间复杂度O(log2n)
// 不稳定
function quickSort(vec, left, right) {
    if (left >= right)
        return;
    let tmp = vec[left];
    let low = left, high = right;
    while (low < high) {
        while (low < high && vec[high] > tmp)
            high--;
        if (low < high)
            vec[low++] = vec[high]; // 注意high不能够动，它的值给出去了其本身就空着了，找一个low那边比它小的赋给它
        while (low < high && vec[low] < tmp)
            low++;
        if (low < high)
            vec[high--] = vec[low]; // 同上
    }
    vec[low] = tmp;
    quickSort(vec, left, low - 1);
    quickSort(vec, low + 1, right);
}

// 选择类排序

/**
 * 直接选择排序
 * @param {Array} vec 待排序数组
 */
// 时间复杂度O(n^2)
// 空间复杂度O(1)
// 不稳定
function selectSort(vec) {
    for (let i = 0; i < vec.length - 1; i++) {
        let idx = i;
        for (let j = i + 1; j < vec.length; j++)
            if (vec[j] < vec[idx])
                idx = j;
        [vec[idx], vec[i]] = [vec[i], vec[idx]];
    }
}

/**
 * 调整堆数组
 * @param {Array} vec 待排序数组
 * @param {number} left 数组中待调整部分闭区间左边界索引
 * @param {number} right 数组中待调整部分闭区间右边界索引
 */
function heapAdjust(vec, left, right) {
    let tmp = vec[left];
    let m = left * 2 + 1;
    while (m <= right) {
        if (m + 1 <= right && vec[m + 1] > vec[m])
            m++;
        if (vec[m] > tmp) {
            vec[left] = vec[m];
            left = m;
            m = left * 2 + 1;
        } else
            break;
    }
    vec[left] = tmp;
}

/**
 * 堆排序
 * @param {Array} vec 待排序数组
 */
// 时间复杂度O(nlog2n)
// 空间复杂度O(1)，在相同时间的算法里是最好的
// 不稳定
function heapSort(vec) {
    for (let i = parseInt((vec.length - 2) / 2); i >= 0; i--)
        heapAdjust(vec, i, vec.length - 1); // 注意这里是从最后一个不为叶节点的节点开始，每次以当前节点为根节点进行调整
    for (let i = vec.length - 1; i >= 1; i--) {
        [vec[0], vec[i]] = [vec[i], vec[0]];
        heapAdjust(vec, 0, i - 1);
    }
}

// 其他排序算法

/**
 * 将数组中两个有序部分归并起来
 * @param {Array} vec 待排序数组
 * @param {number} left 左待归并部分左边界索引
 * @param {number} mid 左右待归并部分分界索引，左部分的右边界索引，右部分的左边界索引-1
 * @param {number} right 右待归并部分右边界索引
 */
function merge(vec, left, mid, right) {
    let lvec = vec.slice(left, mid + 1);
    let rvec = vec.slice(mid + 1, right + 1);
    let li = 0, ri = 0;
    while (li < lvec.length && ri < rvec.length)
        if (lvec[li] <= rvec[ri])
            vec[left++] = lvec[li++];
        else
            vec[left++] = rvec[ri++];
    while (li < lvec.length)
        vec[left++] = lvec[li++];
    while (ri < rvec.length)
        vec[left++] = rvec[ri++];
}

/**
 * 归并排序
 * @param {Array} vec 待排序数组
 * @param {number} left 数组中待归并部分左闭区间边界索引
 * @param {number} right 数组中待归并部分右闭区间边界索引
 */
// 时间复杂度O(nlog2n)
// 空间复杂度O(n)
// 稳定
function mergeSort(vec, left, right) {
    if (left >= right)
        return;
    let mid = parseInt((left + right) / 2);
    mergeSort(vec, left, mid);
    mergeSort(vec, mid + 1, right);
    merge(vec, left, mid, right);
}

/**
 * 基数排序每趟分桶排序
 * @param {Array} vec 待排序数组
 * @param {number} k 从低位开始按照第几位数进行排序
 */
function radix(vec, k) {
    let qvec = new Array();
    for (let i = 0; i < 10; i++)
        qvec.push(new Array());
    let m = Math.pow(10, k - 1);
    vec.forEach(e => {
        qvec[parseInt(e / m) % 10].push(e);
    });
    vec.splice(0);
    qvec.forEach(e => {
        while (e.length > 0)
            vec.push(e.shift());
    });
}

/**
 * 基数排序
 * @param {Array} vec 待排序数组
 * @param {number} k 关键字位数
 */
// 时间复杂度为O(d*(rd+n))，其中d为关键字位数（排序几趟），rd为关键字基个数（0-9），n为元素个数
// 空间复杂度为O(rd)，rd个队列
// 稳定
function radixSort(vec, k) {
    for (let i = 1; i <= k; i++)
        radix(vec, i);
}

/**
 * 双向冒泡排序
 * @param {Array} vec 待排序数组
 */
// 时间复杂度最好O(n)，最坏O(n^2)
// 空间复杂度O(1)
// 稳定
function dbubbleSort(vec) {
    let left = 0, right = vec.length - 1;
    while (left < right) {
        let flag = false;
        for (let i = left; i < right; i++)
            if (vec[i] > vec[i + 1]) {
                [vec[i], vec[i + 1]] = [vec[i + 1], vec[i]];
                flag = true;
            }
        right--;
        for (let i = right; i > left; i--)
            if (vec[i] < vec[i - 1]) {
                [vec[i], vec[i - 1]] = [vec[i - 1], vec[i]];
                flag = true;
            }
        left++;
        if (!flag)
            return;
    }
}

/**
 * 计数排序
 * @param {Array} vec 待排序数组
 * @param {Array} rvec 结果放置数组
 */
// 时间复杂度O(n^2)
// 空间复杂度O(n)
function calcSort(vec, rvec) {
    if (rvec.length != vec.length) {
        rvec.splice(0);
        for (let i = 0; i < vec.length; i++)
            rvec.push(0);
    }
    vec.forEach(e => {
        let index = 0;
        vec.forEach(ee => {
            if (ee < e)
                index++;
        });
        rvec[index] = e;
    });
}

// /*********************************************************************************************************************/

/**
 * 主函数
 */
function main() {
    const len = 10;
    let randvec = genRandVec(len, 0, 100);
    printVec(randvec, "original vec");
    let vec;

    vec = JSON.parse(JSON.stringify(randvec));
    insertSort(vec);
    printVec(vec, "after insert sort");

    vec = JSON.parse(JSON.stringify(randvec));
    binsertSort(vec);
    printVec(vec, "after binary insert sort");

    vec = JSON.parse(JSON.stringify(randvec));
    shellSort(vec, 3, k => { return parseInt(k / 2); });
    printVec(vec, "after shell sort");

    vec = JSON.parse(JSON.stringify(randvec));
    bubbleSort(vec);
    printVec(vec, "after bubble sort");

    vec = JSON.parse(JSON.stringify(randvec));
    quickSort(vec, 0, len - 1);
    printVec(vec, "after quick sort");

    vec = JSON.parse(JSON.stringify(randvec));
    selectSort(vec);
    printVec(vec, "after select sort");

    vec = JSON.parse(JSON.stringify(randvec));
    heapSort(vec);
    printVec(vec, "after heap sort");

    vec = JSON.parse(JSON.stringify(randvec));
    mergeSort(vec, 0, len - 1);
    printVec(vec, "after merge sort");

    vec = JSON.parse(JSON.stringify(randvec));
    radixSort(vec, 3);
    printVec(vec, "after radix sort");

    vec = JSON.parse(JSON.stringify(randvec));
    dbubbleSort(vec);
    printVec(vec, "after dbubble sort");

    vec = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1];
    let rvec = new Array();
    calcSort(vec, rvec);
    printVec(rvec, "after calc sort");
}

main();

File: 数据结构与算法/排序/sort.py
'''

排序算法

'''

from random import randint
from copy import deepcopy
from queue import Queue
from typing import List


# 生成一个随机整数组
def genrandvec(length: int, left: int, right: int) -> List[int]:
    vec = []
    while length > 0:
        vec.append(randint(left, right))
        length -= 1
    return vec


# 输出一个数组
# prefix为输出前缀
def printvec(vec: list, prefix: str):
    print(f"{prefix}:", end="")
    for e in vec:
        print(f" {e}", end="")
    print()


# ###################### 插入类排序算法 ######################


# 直接插入排序
# 时间复杂度最坏为O(n^2)，最好O(n)，平均O(n^2)
# 空间复杂度O(1)
# 是一个稳定的算法
def insertSort(vec: list):
    for i in range(1, len(vec)):
        tmp = vec[i]
        idx = i - 1
        while idx >= 0 and vec[idx] > tmp:
            vec[idx + 1] = vec[idx]
            idx -= 1
        vec[idx + 1] = tmp


# 有序序列中获取待插入索引
def bindex(vec: list, left: int, right: int, target) -> int:
    mid = int((left + right) / 2)
    while left < right:
        if vec[mid] == target:
            return mid + 1
        elif vec[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
        mid = int((left + right) / 2)
    return mid + 1 if vec[mid] <= target else mid


# 折半插入排序
# 时间复杂度最好O(nlog2n)，最坏O(n^2)，平均O(n^2)
# 空间复杂度O(1)
# 是一个稳定的算法
def binsertSort(vec: list):
    for i in range(1, len(vec)):
        index = bindex(vec, 0, i - 1, vec[i])
        tmp = vec[i]
        idx = i - 1
        while idx >= index:
            vec[idx + 1] = vec[idx]
            idx -= 1
        vec[idx + 1] = tmp


# 希尔排序
# 具体时间复杂度和增量选取有关，/2的复杂度为O(n^2)，平均O(nlog2n)
# 空间复杂度O(1)
# 是一个不稳定的算法
def shellSort(vec: list, k: int, kfun):
    while k > 1:
        for i in range(0, k):
            for j in range(i + k, len(vec), k):
                tmp = vec[j]
                idx = j - k
                while idx >= i and vec[idx] > tmp:
                    vec[idx + k] = vec[idx]
                    idx -= k
                vec[idx + k] = tmp
        k = kfun(k)
    for i in range(1, len(vec)):
        tmp = vec[i]
        idx = i - 1
        while idx >= 0 and vec[idx] > tmp:
            vec[idx + 1] = vec[idx]
            idx -= 1
        vec[idx + 1] = tmp


# ###################### 交换类排序算法 ######################


# 冒泡排序
# 时间复杂度最坏为O(n^2)，最好O(n)，平均O(n^2)
# 空间复杂度O(1)
# 稳定
def bubbleSort(vec: list):
    for i in range(len(vec) - 1, 0, -1):
        flag = False
        for j in range(0, i):
            if vec[j] > vec[j + 1]:
                vec[j], vec[j + 1] = vec[j + 1], vec[j]
                flag = True
        if not flag:
            return


# 快速排序
# 时间复杂度最好O(nlog2n)，最坏O(n^2)，平均O(nlog2n)，基本操作次数是所有O(nlog2n)算法里最少的
# 空间复杂度O(log2n)
# 不稳定
def quickSort(vec: list, left: int, right: int):
    if left >= right:
        return
    tmp = vec[left]
    low = left
    high = right
    while low < high:
        while low < high and vec[high] > tmp:
            high -= 1
        if low < high:
            vec[low] = vec[high]
            low += 1
        while low < high and vec[low] < tmp:
            low += 1
        if low < high:
            vec[high] = vec[low]
            high -= 1
    vec[low] = tmp
    quickSort(vec, left, low - 1)
    quickSort(vec, low + 1, right)


# ###################### 选择类排序算法 ######################


# 直接选择排序
def selectSort(vec: list):
    for i in range(0, len(vec) - 1):
        idx = i
        for j in range(i + 1, len(vec)):
            if vec[j] < vec[idx]:
                idx = j
        vec[i], vec[idx] = vec[idx], vec[i]


# 在数组的某个范围中调整根节点
def heapAdjust(vec: list, left: int, right: int):
    tmp = vec[left]
    m = left * 2 + 1
    while m <= right:
        if m + 1 <= right and vec[m + 1] > vec[m]:  # 选一个最大的出来
            m += 1
        if vec[m] > tmp:
            vec[left] = vec[m]
            left = m
            m = left * 2 + 1
        else:
            break
    vec[left] = tmp


# 堆排序
# 时间复杂度O(nlog2n)
# 空间复杂度O(1)，在所有时间O(nlog2n)算法中最好
# 不稳定
# 适合关键字多的场景
def heapSort(vec: list):
    for i in range(int((len(vec) - 2) / 2), -1, -1):
        heapAdjust(vec, i, len(vec) - 1)
    for i in range(len(vec) - 1, 0, -1):
        vec[0], vec[i] = vec[i], vec[0]
        heapAdjust(vec, 0, i - 1)


# ###################### 其他排序算法 ######################


# 合并数组的两个有序部分
def merge(vec: list, left: int, mid: int, right: int):
    a = vec[left:mid + 1]
    b = vec[mid + 1:right + 1]
    ai = 0
    bi = 0
    while ai < len(a) and bi < len(b):
        if a[ai] <= b[bi]:
            vec[left] = a[ai]
            ai += 1
        else:
            vec[left] = b[bi]
            bi += 1
        left += 1
    while ai < len(a):
        vec[left] = a[ai]
        ai += 1
        left += 1
    while bi < len(b):
        vec[left] = b[bi]
        bi += 1
        left += 1


# 二路归并排序
# 时间复杂度O(nlog2n)
# 空间复杂度O(n)
# 稳定
def mergeSort(vec: list, left: int, right: int):
    if left >= right:
        return
    mid = int((left + right) / 2)
    mergeSort(vec, left, mid)
    mergeSort(vec, mid + 1, right)
    merge(vec, left, mid, right)


# 针对某个数位上的数进行的排序
def radix(vec: list, k: int):
    qvec = []
    for i in range(10):
        qvec.append(Queue())
    m = 10**(k - 1)
    for i in vec:
        qvec[int(i / m) % 10].put(i)
    vec.clear()
    for q in qvec:
        while not q.empty():
            vec.append(q.get())


# 基数排序 最低位优先
# 貌似只能排正数，如果有负数的话需要额外九个桶
# 时间复杂度O(d*(rd+n))，其中d为关键字位数（跑几轮排序），rd为关键字基的位数（几个桶），n为关键字个数
# 空间复杂度为O(rd)（桶）
def radixSort(vec: list, k: int):
    for i in range(1, k + 1):
        radix(vec, i)


# 双向冒泡排序
# 时间复杂度最好O(n)，前后扫描两遍，最坏O(n^2)，一直扫描到两边指针碰头
# 空间复杂度O(1)，没有随规模变化的额外空间
# 稳定
def dbubbleSort(vec: list):
    left = 0
    right = len(vec) - 1
    while left < right:
        flag = False
        for i in range(left, right):
            if vec[i] > vec[i + 1]:
                vec[i], vec[i + 1] = vec[i + 1], vec[i]
                flag = True
        right -= 1
        for i in range(right, left, -1):
            if vec[i] < vec[i - 1]:
                vec[i], vec[i - 1] = vec[i - 1], vec[i]
                flag = True
        left += 1
        if not flag:
            return


# 计数排序
# 要求待排序数组内每个元素都不相同
# 时间复杂度O(n^2)，因为对于每个元素都要扫描一遍数组
# 空间复杂度O(n)，因为有额外的辅助数组
# 谈不上稳不稳定，因为它根本就没有相同的元素
def calcSort(vec: list, rvec: list):
    for i in vec:
        idx = 0
        for j in vec:
            if j < i:
                idx += 1
        rvec[idx] = i


if __name__ == "__main__":
    length = 10
    randvec = genrandvec(length, 0, 100)
    printvec(randvec, "original vec")

    vec = deepcopy(randvec)
    insertSort(vec)
    printvec(vec, "after insert sort")

    vec = deepcopy(randvec)
    binsertSort(vec)
    printvec(vec, "after binary insert sort")

    vec = deepcopy(randvec)
    shellSort(vec, 5, lambda x: x // 2)
    printvec(vec, "after shell sort")

    vec = deepcopy(randvec)
    bubbleSort(vec)
    printvec(vec, "after bubble sort")

    vec = deepcopy(randvec)
    quickSort(vec, 0, length - 1)
    printvec(vec, "after quick sort")

    vec = deepcopy(randvec)
    selectSort(vec)
    printvec(vec, "after select sort")

    vec = deepcopy(randvec)
    heapSort(vec)
    printvec(vec, "after heap sort")

    vec = deepcopy(randvec)
    mergeSort(vec, 0, length - 1)
    printvec(vec, "after binary merge sort")

    vec = deepcopy(randvec)
    radixSort(vec, 3)
    printvec(vec, "after radix sort")

    vec = deepcopy(randvec)
    dbubbleSort(vec)
    printvec(vec, "after dbubble sort")

    vec = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
    rvec = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    calcSort(vec, rvec)
    printvec(rvec, "after calc sort")
